source/extensions/quic_listeners/quiche/platform/quic_test_output_impl.cc

```diff
@@ -0,0 +1,71 @@
+// NOLINT(namespace-envoy)
+
+// This file is part of the QUICHE platform implementation, and is not to be
+// consumed or referenced directly by other Envoy code. It serves purely as a
+// porting layer for QUICHE.
+
+#include "extensions/quic_listeners/quiche/platform/quic_test_output_impl.h"
+
+#include <cstdlib>
+
+#include "common/filesystem/filesystem_impl.h"
+
+#include "absl/time/clock.h"
+#include "absl/time/time.h"
+#include "fmt/printf.h"
+#include "gtest/gtest.h"
+#include "quiche/quic/platform/api/quic_logging.h"
+
+namespace quic {
+
+void QuicRecordTestOutputToFile(const std::string& filename, QuicStringPiece data) {
+  const char* output_dir_env = std::getenv("QUIC_TEST_OUTPUT_DIR");
+  if (output_dir_env == nullptr) {
+    QUIC_LOG(WARNING) << "Could not save test output since QUIC_TEST_OUTPUT_DIR is not set";
+    return;
+  }
+
+  std::string output_dir = output_dir_env;
+  if (output_dir.empty()) {
+    QUIC_LOG(WARNING) << "Could not save test output since QUIC_TEST_OUTPUT_DIR is empty";
+    return;
+  }
+
+  if (output_dir.back() != '/') {
+    output_dir += '/';
+  }
+
+  Envoy::Filesystem::InstanceImpl file_system;
```

Comment:
It would be nice if the Filesystem::Instance could be obtained from Api::Api::fileSystem(), but I can't think of a way to do that, since we don't have access to the Api::Api instance here, and there's no good way to pass it down explicitly.

What you have here looks like best immediate solution to me, but leaving this comment in case Envoy folks have any other ideas. I do wonder if this will come up in other places as well though, e.g. if other platform impls need to make use of functionality abstracted under Api::Api. Maybe we need a singleton (ick) or other means to fetch the Api::Api when it can't be passed explicitly?

Reply:
I'd love @mattklein123 take on this one.
I had a similar issue with something the server owned lately and I reaaaally wanted to make it a singleton too, but then again I love singletons about as much as Matt and Josh dislike them.

Reply:
The code in this PR is purely for tests, IMHO, making test code more testable via a mock is going a little bit too far. Does it make any sense?

Reply:
Fair enough, I'm fine with not using a mock and checking for the file directly.
I do think we will need a solution to the general Api::Api issue, but maybe we should deal with that separately from this PR? If we do that, we'll need to remember to circle back and update quic_test_output_impl.cc to not hardcode FilesystemImpl later on.

Reply:
I think we are going to have to end up plumbing an `Api` into this platform abstraction somehow. Can the platform abstraction have shared state that is initialized? If we have that it should be quite easy to pass in an `Api`. cc @jmarantz

Reply:
@mpwarres Agree we need a solution for the general Api issue, I think we can circle back when we have a concrete issue related to it. I still think quic_test_output_impl.cc can continue to write directly to real files. I found that test function [TestEnvironment::writeStringToFileForTest](https://github.com/envoyproxy/envoy/blob/master/test/test_common/environment.cc#L295) is already doing something similar.

@mattklein123 The code in this PR is test only, all it does is to allow some unit tests to save some test output to somewhere(in this case, a file). We have a test of this test-code in quic_platform_test.cc(QuicPlatformTest.QuicTestOutput), I think the point of plumbing Api into here is to test this test-code _better_, which seems a little overkill?

@mattklein123 Most platform abstractions, including this one, do not have shared states. I won't be surprised if some of them end up needing a shared Api, I think we can deal with that when the need arises.

Reply:
If this is only used in test code, can we put it in the test directory to make that clear? Can you have a test platform split out?

Reply:
@mattklein123 I have tried moving it to the test directory previously, but gave up due to some complications. The issue is quic_test_output.h, which is part of upstream QUICHE, and built in cc_library "quic_platform_base" in bazel/external/quiche.BUILD, wants to include 'extensions/quic_listeners/quiche/platform/quic_test_output_impl.h', this path does not work if quic_test_output_impl.h is in the test directory.

Reply:
Re: @mattklein123's  earlier [question](https://github.com/envoyproxy/envoy/pull/6058/files#r260978895), we could add shared state to the platform implementation that is initialized early on via an Init() call, which IIUC would amount to a singleton maintained within the QUICHE platform impl. If that seems reasonable, we can add that, though I think our preference would be to do it separate from this PR, if that's ok.